File: uniffi/src/message/groups/first_seen_replay_couples.rs
use cesrox_core::error::CesrResult;
use cesrox_core::{Matter, Group};
pub use cesrox_core::message::groups::{
    FirstSeenReplayCouple,
    FirstSeenReplayCouples,
};

pub fn first_seen_replay_couple_create(firner: Matter, dater: Matter) -> FirstSeenReplayCouple {
    FirstSeenReplayCouple::new(firner, dater)
}

pub fn first_seen_replay_couples_create(value: Vec<FirstSeenReplayCouple>) -> FirstSeenReplayCouples {
    FirstSeenReplayCouples::new(value)
}

pub fn first_seen_replay_couples_qb64(first_seen_replay_couples: &FirstSeenReplayCouples) -> CesrResult<String> {
    first_seen_replay_couples.qb64()
}

pub fn first_seen_replay_couples_qb64b(first_seen_replay_couples: &FirstSeenReplayCouples) -> CesrResult<Vec<u8>> {
    first_seen_replay_couples.qb64b()
}

pub fn first_seen_replay_couples_qb2(first_seen_replay_couples: &FirstSeenReplayCouples) -> CesrResult<Vec<u8>> {
    first_seen_replay_couples.qb2()
}

File: wrappers/kotlin/lib/src/main/kotlin/cesrox/groups/FirstSeenReplayCouples.kt
package cesrox.groups

import cesrox.primitives.Seqner
import cesrox.primitives.Dater
import uniffi.cesrox.*
import uniffi.cesrox.FirstSeenReplayCouple as FirstSeenReplayCoupleNative
import uniffi.cesrox.FirstSeenReplayCouples as FirstSeenReplayCouplesNative

class FirstSeenReplayCouple {
    val value: FirstSeenReplayCoupleNative

    constructor(firner: Seqner, dater: Dater) {
        value = firstSeenReplayCoupleCreate(firner.value, dater.value)
    }

    constructor(nativeValue: FirstSeenReplayCoupleNative) {
        value = nativeValue
    }

    val firner get() = Seqner(value.firner)

    val dater get() = Dater(value.dater)
}

class FirstSeenReplayCouples {
    val value: FirstSeenReplayCouplesNative

    constructor(couples: List<FirstSeenReplayCouple>) {
        value = firstSeenReplayCouplesCreate(couples.map { it.value })
    }

    constructor(couples: FirstSeenReplayCouplesNative) {
        value = couples
    }

    fun get(index: Int): FirstSeenReplayCouple {
        return FirstSeenReplayCouple(value.value[index])
    }

    fun qb2(): List<UByte> {
        return firstSeenReplayCouplesQb2(value)
    }

    fun qb64b(): List<UByte> {
        return firstSeenReplayCouplesQb64b(value)
    }

    fun qb64(): String {
        return firstSeenReplayCouplesQb64(value)
    }
}